cpp/cmake_modules/ThirdpartyToolchain.cmake

```diff
@@ -4845,6 +4845,14 @@ macro(build_opentelemetry)
       version)
   set(OPENTELEMETRY_BUILD_BYPRODUCTS)
   set(OPENTELEMETRY_LIBRARIES)
+  set(OPENTELEMETRY_CXX_FLAGS "${EP_CXX_FLAGS}")
+  # GH-46508: Supress warnings that won't be demoted by -Wno-error in recent Clang.
+  if((CMAKE_CXX_COMPILER_ID STREQUAL "Clang" AND
+      CMAKE_CXX_COMPILER_VERSION VERSION_GREATER_EQUAL "19.1.0") OR
+     (CMAKE_CXX_COMPILER_ID STREQUAL "AppleClang" AND
+      CMAKE_CXX_COMPILER_VERSION VERSION_GREATER_EQUAL "17.0.0"))
+    string(APPEND OPENTELEMETRY_CXX_FLAGS " -Wno-error=missing-template-arg-list-after-template-kw")
+  endif()
```

Comment:
Can we always omit `-Werror` in OpenTelemetry C++? In general, we don't want to enable `-Werror` for bundled dependencies.

Reply:
Thank you @kou . But omitting `-Werror` doesn't suppress this error. I'm not exactly sure how this warning as error works but here are some godbolt experiments showing only `-Wno-error=missing-template-arg-list-after-template-kw` would work.
No `-Werror` (error): https://godbolt.org/z/8bMdPEaoE
Solely `-Wno-error` (error): https://godbolt.org/z/f7KW3MsP5
Solely `-Wno-error=missing-template-arg-list-after-template-kw` (ok): https://godbolt.org/z/bf3Gds85n

Reply:
Oh, sorry!

BTW, does it still happen with the latest OpenTelemetry C++?

Reply:
Hmm, didn't try. But I can have a try since this shouldn't be much work.

Reply:
The latest opentelemetry-cpp doesn't have this error, only that we need a small tweak to override a virtual method that has been made pure virtual in recent version - not a big deal though.

I made the change in [`8a718df` (#46509)](https://github.com/apache/arrow/pull/46509/commits/8a718dfa328ee847dbb44a8b30ae8b56355b7e71) , which has been verified to be working w/o the previous changes (the compiler options). I'm keeping both fixes temporarily for you to choose which is preferable and will remove the other once we settled on one. I personally think upgrading to latest makes more sense.

@kou PTAL, thanks.

Reply:
Thanks for reminding. I just reviewed it.